File: server/src/main/scala/producer/message_generator/MessageGenerator.scala
package producer.message_generator

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import _root_.producer.message_generator.key_generator.KeyGenerator
import _root_.producer.message_generator.value_generator.ValueGenerator
import _root_.producer.message_generator.properties_generator.PropertiesGenerator
import _root_.producer.message_generator.event_time_generator.EventTimeGenerator
import _root_.producer.message_generator.deliver_at_generator.DeliverAtGenerator
import _root_.producer.message_generator.deliver_after_generator.DeliverAfterGenerator
import _root_.producer.message_generator.ordering_key_generator.OrderingKeyGenerator
import _root_.producer.message_generator.sequence_id_generator.SequenceIdGenerator

case class MessageGenerator(
    keyGenerator: Option[KeyGenerator],
    valueGenerator: Option[ValueGenerator],
    propertiesGenerator: Option[PropertiesGenerator],
    eventTimeGenerator: Option[EventTimeGenerator],
    deliverAtGenerator: Option[DeliverAtGenerator],
    deliverAfterGenerator: Option[DeliverAfterGenerator],
    orderingKeyGenerator: Option[OrderingKeyGenerator],
    sequenceIdGenerator: Option[SequenceIdGenerator]
)

object MessageGenerator:
    def fromPb(v: pb.MessageGenerator): MessageGenerator =
        MessageGenerator(
            keyGenerator = v.keyGenerator.map(KeyGenerator.fromPb),
            valueGenerator = v.valueGenerator.map(ValueGenerator.fromPb),
            propertiesGenerator = v.propertiesGenerator.map(PropertiesGenerator.fromPb),
            eventTimeGenerator = v.eventTimeGenerator.map(EventTimeGenerator.fromPb),
            deliverAtGenerator = v.deliverAtGenerator.map(DeliverAtGenerator.fromPb),
            deliverAfterGenerator = v.deliverAfterGenerator.map(DeliverAfterGenerator.fromPb),
            orderingKeyGenerator = v.orderingKeyGenerator.map(OrderingKeyGenerator.fromPb),
            sequenceIdGenerator = v.sequenceIdGenerator.map(SequenceIdGenerator.fromPb)
        )

    def toPb(v: MessageGenerator): pb.MessageGenerator =
        pb.MessageGenerator(
            keyGenerator = v.keyGenerator.map(KeyGenerator.toPb),
            valueGenerator = v.valueGenerator.map(ValueGenerator.toPb),
            propertiesGenerator = v.propertiesGenerator.map(PropertiesGenerator.toPb),
            eventTimeGenerator = v.eventTimeGenerator.map(EventTimeGenerator.toPb),
            deliverAtGenerator = v.deliverAtGenerator.map(DeliverAtGenerator.toPb),
            deliverAfterGenerator = v.deliverAfterGenerator.map(DeliverAfterGenerator.toPb),
            orderingKeyGenerator = v.orderingKeyGenerator.map(OrderingKeyGenerator.toPb),
            sequenceIdGenerator = v.sequenceIdGenerator.map(SequenceIdGenerator.toPb)
        )

File: ui/components/ui/ProducerSession/message-generator/message-generator.ts
import * as pb from '../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { DeliverAfterGenerator, deliverAfterGeneratorFromPb, deliverAfterGeneratorToPb } from './deliver-after-generator/deliver-after-generator';
import { DeliverAtGenerator, deliverAtGeneratorFromPb, deliverAtGeneratorToPb } from './deliver-at-generator/deliver-at-generator';
import { EventTimeGenerator, eventTimeGeneratorFromPb, eventTimeGeneratorToPb } from './event-time-generator/event-time-generator';
import { KeyGenerator, keyGeneratorFromPb, keyGeneratorToPb } from './key-generator/key-generator';
import { OrderingKeyGenerator, orderingKeyGeneratorFromPb, orderingKeyGeneratorToPb } from './ordering-key-generator/ordering-key-generator';
import { PropertiesGenerator, propertiesGeneratorFromPb, propertiesGeneratorToPb } from './properties-generator/properties-generator';
import { SequenceIdGenerator, sequenceIdGeneratorFromPb, sequenceIdGeneratorToPb } from './sequence-id-generator/sequence-id-generator';
import { ValueGenerator, valueGeneratorFromPb, valueGeneratorToPb } from './value-generator/value-generator';

export type MessageGenerator = {
  type: 'message-generator',
  keyGenerator: KeyGenerator | undefined,
  valueGenerator: ValueGenerator | undefined,
  propertiesGenerator: PropertiesGenerator | undefined,
  eventTimeGenerator: EventTimeGenerator | undefined,
  deliverAtGenerator: DeliverAtGenerator | undefined,
  deliverAfterGenerator: DeliverAfterGenerator | undefined,
  orderingKeyGenerator: OrderingKeyGenerator | undefined,
  sequenceIdGenerator: SequenceIdGenerator | undefined
};

export function messageGeneratorFromPb(v: pb.MessageGenerator): MessageGenerator {
  return {
    type: 'message-generator',
    keyGenerator: v.hasKeyGenerator() ? keyGeneratorFromPb(v.getKeyGenerator()!) : undefined,
    valueGenerator: v.hasValueGenerator() ? valueGeneratorFromPb(v.getValueGenerator()!) : undefined,
    propertiesGenerator: v.hasPropertiesGenerator() ? propertiesGeneratorFromPb(v.getPropertiesGenerator()!) : undefined,
    eventTimeGenerator: v.hasEventTimeGenerator() ? eventTimeGeneratorFromPb(v.getEventTimeGenerator()!) : undefined,
    deliverAtGenerator: v.hasDeliverAtGenerator() ? deliverAtGeneratorFromPb(v.getDeliverAtGenerator()!) : undefined,
    deliverAfterGenerator: v.hasDeliverAfterGenerator() ? deliverAfterGeneratorFromPb(v.getDeliverAfterGenerator()!) : undefined,
    orderingKeyGenerator: v.hasOrderingKeyGenerator() ? orderingKeyGeneratorFromPb(v.getOrderingKeyGenerator()!) : undefined,
    sequenceIdGenerator: v.hasSequenceIdGenerator() ? sequenceIdGeneratorFromPb(v.getSequenceIdGenerator()!) : undefined
  };
}

export function messageGeneratorToPb(v: MessageGenerator): pb.MessageGenerator {
  const generatorPb = new pb.MessageGenerator();
  if (v.keyGenerator !== undefined) {
    generatorPb.setKeyGenerator(keyGeneratorToPb(v.keyGenerator));
  }
  if (v.valueGenerator !== undefined) {
    generatorPb.setValueGenerator(valueGeneratorToPb(v.valueGenerator));
  }
  if (v.propertiesGenerator !== undefined) {
    generatorPb.setPropertiesGenerator(propertiesGeneratorToPb(v.propertiesGenerator));
  }
  if (v.eventTimeGenerator !== undefined) {
    generatorPb.setEventTimeGenerator(eventTimeGeneratorToPb(v.eventTimeGenerator));
  }
  if (v.deliverAtGenerator !== undefined) {
    generatorPb.setDeliverAtGenerator(deliverAtGeneratorToPb(v.deliverAtGenerator));
  }
  if (v.deliverAfterGenerator !== undefined) {
    generatorPb.setDeliverAfterGenerator(deliverAfterGeneratorToPb(v.deliverAfterGenerator));
  }
  if (v.orderingKeyGenerator !== undefined) {
    generatorPb.setOrderingKeyGenerator(orderingKeyGeneratorToPb(v.orderingKeyGenerator));
  }
  if (v.sequenceIdGenerator !== undefined) {
    generatorPb.setSequenceIdGenerator(sequenceIdGeneratorToPb(v.sequenceIdGenerator));
  }

  return generatorPb;
}
